### PR TITLE
Update tutorials using PL-QChem functions to build the electronic Hamiltonian

### PR DESCRIPTION
**Pull request by agran2018**
**Summary:**
This PR updates four tutorials with the PL-QChem refactored functions use to build the electronic Hamiltonian of molecules.

Updated tutorials:

1- "_Quantum Chemistry with PennyLane_" implemented in `tutorial_quantum_chemistry.py`
2- "_A brief overview of VQE_" implemented in `tutorial_vqe.py`
3- "_VQE with parallel QPUs on Rigetti Forest_" implemented in `tutorial_vqe_parallel.py`
4- "_Accelerating VQEs with quantum natural gradient_" implemented in `tutorial_vqe_qng.py`

**Relevant references:**
Included in the tutorials

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: demonstrations/tutorial_quantum_chemistry.py [hunk 1]
```diff
@@ -9,21 +9,21 @@
 
 In quantum chemistry and materials science, the term *electronic structure methods* encompasses
 the approximations used to find the many-electron wave function of polyatomic systems.
-Electronic structure methods rely on the Born-Oppenheimer approximation
-:ref:`[1]<qchem_references>`, which allows to write the electronic Hamiltonian of the molecule as
+Electronic structure methods rely on the Born-Oppenheimer approximation [[#kohanoff2006]_]
+, which allows to write the electronic Hamiltonian of the molecule as
```
Comment (ixfoduap) [inline demonstrations/tutorial_quantum_chemistry.py:13]:
```suggestion
, which allows us to write the electronic Hamiltonian of the molecule as

```

File: demonstrations/tutorial_quantum_chemistry.py [hunk 2]
```diff
@@ -195,24 +200,25 @@
 # Let's partition the HF orbitals to define an active space of four electrons in four active
 # orbitals:
 
-d_occ_indices, active_indices = qml.qchem.active_space(name, hf_data,
-                                                       n_active_electrons=4,
-                                                       n_active_orbitals=4)
+electrons = 10
+orbitals = 7
+core, active = qchem.active_space(electrons, orbitals, active_electrons=4, active_orbitals=4)
```
Comment (ixfoduap) [inline demonstrations/tutorial_quantum_chemistry.py:205]: Love this!

File: demonstrations/tutorial_quantum_chemistry.py [hunk 3]
```diff
@@ -195,24 +200,25 @@
...
-# Notice that calling the :func:`~.active_space` function without specifying an active
-# space results in no doubly-occupied orbitals---*all* molecular orbitals are considered to be active.
+# Notice that calling the :func:`~.pennylane_qchem.qchem.active_space` function without
+# specifying an active space results in no core orbitals---*all* molecular orbitals are
+# considered to be active.
 
-no_d_occ, all_active = qml.qchem.active_space(name, hf_data)
-print("List of doubly-occupied molecular orbitals: {:}".format(no_d_occ))
-print("List of active molecular molecular orbitals: {:}".format(all_active))
+no_core, all_active = qchem.active_space(electrons, orbitals)
```
Comment (ixfoduap) [inline demonstrations/tutorial_quantum_chemistry.py:219]: So `no_core` stands for "no core"?

Comment (ixfoduap) [reply demonstrations/tutorial_quantum_chemistry.py:219]: I find this a bit confusing. Worth revising

File: demonstrations/tutorial_quantum_chemistry.py [hunk 4]
```diff
@@ -237,41 +243,35 @@
...
-	name,
-	'h2o.xyz',
-	charge,
-	multiplicity,
-	basis_set,
-	qc_package='pyscf',
-	n_active_electrons=4,
-	n_active_orbitals=4,
-	mapping='jordan_wigner'
+H, qubits = qchem.molecular_hamiltonian(
```
Comment (ixfoduap) [inline demonstrations/tutorial_quantum_chemistry.py:260]: Do we want to package the inputs from `name` to `qc_package` as a separate variable, e.g. `main_input` so we can call

`molecular_hamiltonian(main_input, active_electrons=4, active_orbitals=4, mapping='jordan_wigner')` ?

And there has to be a better name than `main_input`, it's just the first that came to my mind

File: demonstrations/tutorial_vqe.py [hunk 5]
```diff
@@ -191,8 +192,7 @@ def circuit(params, wires):
     conv = np.abs(energy - prev_energy)
 
     if n % 20 == 0:
-        print('Iteration = {:},  Ground-state energy = {:.8f} Ha,  Convergence parameter = {'
-              ':.8f} Ha'.format(n, energy, conv))
+        print('Iteration = {:},  Energy = {:.8f} Ha'.format(n, energy))
```
Comment (ixfoduap) [inline demonstrations/tutorial_vqe.py:195]: Love this!

File: demonstrations/tutorial_vqe_parallel.py [hunk 6]
```diff
@@ -56,7 +56,7 @@
 # follow a standard format for specifying the geometry of a molecule and can be downloaded as a
 # Zip from :download:`here <../demonstrations/vqe_parallel/vqe_parallel.zip>`.
 
-data = {  # keys: atomic separations (in Angstroms), values: corresponding files
+data = {  # keys: atomic separations (in Angstrom), values: corresponding files
```
Comment (ixfoduap) [inline demonstrations/tutorial_vqe_parallel.py:59]: Not sure about this. We say "in metres", in "inches", so why not "in Angstroms"?

File: demonstrations/tutorial_vqe_qng.py [hunk 7]
```diff
@@ -374,10 +362,9 @@ def ansatz(params, wires=[0, 1, 2, 3]):
     energy = cost(params)
     conv = np.abs(energy - prev_energy)
 
-    if n % 20 == 0:
+    if n % 4 == 0:
```
Comment (ixfoduap) [inline demonstrations/tutorial_vqe_qng.py:365]: Why this change? Having a very long printout in a tutorial can break the flow for the reader

File: demonstrations/tutorial_quantum_chemistry.py (hunk 4 above)
Comment (ixfoduap) [reply demonstrations/tutorial_quantum_chemistry.py:260]: I guess `geometry` is a good way of packaging the inputs?

File: demonstrations/tutorial_quantum_chemistry.py [hunk 8]
```diff
@@ -38,9 +38,9 @@
 
 Importing the molecular structure
 ---------------------------------
-The first step is to import PennyLane.
+The first step is to import the QChem package from PennyLane.
 """
-import pennylane as qml
+from pennylane import qchem
```
Comment (josh146) [inline demonstrations/tutorial_quantum_chemistry.py:43]: What is the reason for this change?

File: demonstrations/tutorial_quantum_chemistry.py (hunk 4 above)
Comment (josh146) [reply demonstrations/tutorial_quantum_chemistry.py:260]: What would `main_input` be here? If a dictionary, that I would say no --- it is more backwards compatible and easier to document by having them as arguments with defaults

**Review by josh146 — COMMENTED** [review]
Thanks @agran2018! I'm changing the base branch to `dev`; the `dev` branch uses GitHub for the requirements, whereas master uses the pip versions.

I might also remove myself as a reviewer :) But feel free to ask if you have any specific questions.

File: demonstrations/tutorial_quantum_chemistry.py (hunk 1 above)
Comment (trbromley) [reply demonstrations/tutorial_quantum_chemistry.py:13]: image
I think the extra square brackets are not needed?

Comment (trbromley) [reply demonstrations/tutorial_quantum_chemistry.py:13]: (same for other references below)

File: demonstrations/tutorial_quantum_chemistry.py [hunk 9]
```diff
@@ -82,19 +83,20 @@
...
 # and `Coupled Cluster (CC) <https://en.wikipedia.org/wiki/Coupled_cluster>`__ methods among
-# others :ref:`[2]<qchem_references>`.
+# others [[#jensenbook]_].
 #
-# Before launching the HF calculation using the function :func:`meanfield_data`, we
-# need to specify a string to label the molecule and the net charge of the molecule. In this
-# example we choose ``'water'`` as the string. On the other hand, although positively or
-# negatively charged molecules can be simulated, we choose a neutral system.
+# Before launching the HF calculation using the function
+# :func:`~.pennylane_qchem.qchem.meanfield`, we need to specify a string to label the
```
Comment (trbromley) [inline demonstrations/tutorial_quantum_chemistry.py:89]: This link wouldn't work :thinking:

File: demonstrations/tutorial_quantum_chemistry.py [hunk 10]
```diff
@@ -136,30 +138,33 @@
 basis_set = 'sto-3g'
 
 ##############################################################################
-# Finally, we can call the function :func:`~.meanfield_data` to launch the mean field
-# calculation. At present, the quantum chemistry packages `PySCF
+# Finally, we can call the function :func:`~.pennylane_qchem.qchem.meanfield` to launch
```
Comment (trbromley) [inline demonstrations/tutorial_quantum_chemistry.py:141]: Link wouldn't work

File: demonstrations/tutorial_quantum_chemistry.py (hunk 3 above)
Comment (trbromley) [reply demonstrations/tutorial_quantum_chemistry.py:219]: Could we just call them `core` and `active` as before?

File: demonstrations/tutorial_quantum_chemistry.py [hunk 11]
```diff
@@ -237,41 +243,35 @@
...
-	docc_mo_indices=d_occ_indices,
-	active_mo_indices=active_indices
-)
+qubit_hamiltonian = qchem.decompose(hf_file, mapping="jordan_wigner", core=core, active=active)
 print("Electronic Hamiltonian of the water molecule represented in the Pauli basis")
 print(qubit_hamiltonian)
 
 ##############################################################################
-# Finally, the :func:`~.generate_hamiltonian`
+# Finally, the :func:`~.pennylane_qchem.qchem.molecular_hamiltonian`
```
Comment (trbromley) [inline demonstrations/tutorial_quantum_chemistry.py:254]: Couldn't get link to work

File: demonstrations/tutorial_vqe.py [hunk 12]
```diff
@@ -7,7 +7,7 @@
         variational quantum eigensolver algorithm in PennyLane.
     :property="og:image": https://pennylane.ai/qml/_images/pes_h2.png
 
-The Variational Quantum Eigensolver (VQE) :ref:`[1, 2]<vqe_references>` is a flagship algorithm for
+The Variational Quantum Eigensolver (VQE) [[#peruzzo2014]_, [#yudong2019]_] is a flagship algorithm for
```
Comment (trbromley) [inline demonstrations/tutorial_vqe.py:10]: Also references in this script don't need the extra brackets

File: demonstrations/tutorial_vqe.py [hunk 13]
```diff
@@ -79,31 +80,31 @@
 ##############################################################################
 # At this stage, to compute the molecule's Hamiltonian in the Pauli basis, several
 # calculations need to be performed. With PennyLane, these can all be done in a
-# single line by calling the function :func:`~.generate_hamiltonian`. The first input to
-# the function is a string denoting the name of the molecule, which will determine the name given
-# to the saved files that are produced during the calculations:
+# single line by calling the function :func:`~.pennylane_qchem.qchem.molecular_hamiltonian`.
```
Comment (trbromley) [inline demonstrations/tutorial_vqe.py:83]: Trouble with link

File: demonstrations/tutorial_vqe_parallel.py [hunk 14]
```diff
@@ -45,7 +45,7 @@
 # the bond length between the hydrogen atoms.
 #
 # Each inter-atomic distance results in a different qubit Hamiltonian. To find the corresponding
-# Hamiltonian, we use the :func:`~.pennylane_qchem.qchem.generate_hamiltonian` function of the
+# Hamiltonian, we use the :func:`~.pennylane_qchem.qchem.molecular_hamiltonian` function of the
```
Comment (trbromley) [inline demonstrations/tutorial_vqe_parallel.py:48]: Also trouble with link

File: demonstrations/tutorial_vqe_qng.py [hunk 15]
```diff
@@ -256,26 +256,15 @@ def circuit(params, wires=0):
 # (2) Hydrogen VQE Example
 # ------------------------
 #
-# To construct our system Hamiltonian, we call the function :func:`~.generate_hamiltonian`.
+# To construct our system Hamiltonian, we call the function
+# :func:`~.pennylane_qchem.qchem.molecular_hamiltonian`.
```
Comment (trbromley) [inline demonstrations/tutorial_vqe_qng.py:260]: Also trouble with link

File: demonstrations/tutorial_vqe_qng.py (hunk 7 above)
Comment (trbromley) [reply demonstrations/tutorial_vqe_qng.py:365]: image
It looks ok

File: demonstrations/tutorial_quantum_chemistry.py (hunk 9 above)
Comment (josh146) [reply demonstrations/tutorial_quantum_chemistry.py:89]: Ah, this is because the intersphinx mapping in `conf.py` is still linking to stable, not latest.

File: demonstrations/tutorial_quantum_chemistry.py (hunk 1 above)
Comment (josh146) [reply demonstrations/tutorial_quantum_chemistry.py:13]: Yep 👍 This comes from the version of sphinx you might be using.

* Sphinx 1.X automatically includes square brackets when you add a reference `[#kohanoff2006]_`

* Sphinx 2.X doesn't; you have to add the square brackets yourself: `[[#kohanoff2006]_]`.

On CircleCI, we pin Sphinx to version 1.8.5, hence the extra brackets aren't needed.

Comment (agran2018) [reply demonstrations/tutorial_quantum_chemistry.py:13]: Thanks @trbromley and @josh146

File: demonstrations/tutorial_quantum_chemistry.py (hunk 8 above)
Comment (agran2018) [reply demonstrations/tutorial_quantum_chemistry.py:43]: Since no core functionalities of PennyLane are used in this tutorial this makes the calling statements shorter. If there are reasons to keep the `qml` prefix, let me know, have no problem with that.

File: demonstrations/tutorial_quantum_chemistry.py (hunk 9 above)
Comment (agran2018) [reply demonstrations/tutorial_quantum_chemistry.py:89]: Yes.

File: demonstrations/tutorial_quantum_chemistry.py (hunk 10 above)
Comment (agran2018) [reply demonstrations/tutorial_quantum_chemistry.py:141]: The same reason as mentioned above @trbromley

File: demonstrations/tutorial_quantum_chemistry.py (hunk 3 above)
Comment (agran2018) [reply demonstrations/tutorial_quantum_chemistry.py:219]: `no_core` stands for "no core orbitals". I could name it `no_core_orbitals` (?)

We can not name it `core` and `active` since we use this variables below to decompose the hamiltonian.
```
qubit_hamiltonian = qchem.decompose(hf_file, mapping="jordan_wigner", core=core, active=active)
```
This is just to show that the user has full flexibility to define different active spaces.

File: demonstrations/tutorial_quantum_chemistry.py (hunk 4 above)
Comment (agran2018) [reply demonstrations/tutorial_quantum_chemistry.py:260]: I agree with @josh146 . 
Another possibility here is to shorten the call statement by using the defaults for `charge`, `mult`, `basis`, `package` and only pass `active_electrons` and `active_orbitals`. However, I think that being explicit in this tutorial with all keyword arguments is pertinent and informative to the user.

File: demonstrations/tutorial_vqe_parallel.py (hunk 6 above)
Comment (agran2018) [reply demonstrations/tutorial_vqe_parallel.py:59]: I am use to seeing "Angstrom", the same for "Bohr" when things are in atomic units. But not 100 % sure. Will change that.

File: demonstrations/tutorial_quantum_chemistry.py (hunk 8 above)
Comment (josh146) [reply demonstrations/tutorial_quantum_chemistry.py:44]: All good 👍 I tend to prefer using `qml.qchem.etc`, but that is just personal preference.

File: demonstrations/tutorial_quantum_chemistry.py (hunk 3 above)
Comment (josh146) [reply demonstrations/tutorial_quantum_chemistry.py:220]: I like `no_core_orbitals`.

**Review by josh146 — APPROVED** [review]
Looks good to me, thanks @agran2018!

This PR is a high priority PR to merge, as it is blocking the `dev` branch of the QML repo from building for other PRs, and required for the upcoming PL release.. @ixfoduap if you could prioritize reviewing so that we could merge this in, I would greatly appreciate it.

File: demonstrations/tutorial_vqe_qng.py [hunk 16]
```diff
@@ -386,6 +373,7 @@ def ansatz(params, wires=[0, 1, 2, 3]):
     qngd_cost.append(energy)
     prev_energy = energy
 
+print()
 print("Final convergence parameter = {:.8f} Ha".format(conv))
```
Comment (josh146) [inline demonstrations/tutorial_vqe_qng.py:377]:
```suggestion
print("\nFinal convergence parameter = {:.8f} Ha".format(conv))
```
